### ISSUE TITLE
less escaping seems to be broken

### ISSUE THREAD

**Issue by rickhuizinga** [ngbp/ngbp#173]
There is a problem with grunt-recess in that less escaping seems to be broken.  See: https://github.com/twitter/recess/issues/59

In summary, using the .less file below

```
.carousel-inner {
  width: 83%;
  width: ~"calc(100% - 71px)";
  width: ~"-webkit-calc(100% - 71px)";
  width: -moz-calc(~"100% - 71px");
  width: -o-calc(~"100% - 71px");
}
```

results in css:

```
.carousel-inner {
  width: 83%;
  width: calc(29%);
  width: -webkit-calc(29%);
  width: -moz-calc(29%);
  width: -o-calc(29%);
}
```

less should however "pass through" all the parts escaped by ~""

This is the wrong place to request a fix for this bug, but was there any reason behind the design choice of using grunt-recess vs. grunt-contrib-less?  I'm thinking of switching my local copy of ng-boilerplate to use grunt-contrib-less and submitting a PR.

**Comment by alex6o** [comment]
same problem here with the escape syntax (grunt-recess#0.4.0); it would be great if you could share your implementation of contrib-less :)

**Comment by rickhuizinga** [comment]
I've switched my working copy of ng-boilerplate to use grunt-contrib-less rather than grunt-recess, and escaping now works.  Statements such as these now are properly compiled into the CSS:

max-width: ~"calc(100% - @{lessVariable})";

I can submit a PR if the project maintainer agrees to switch to grunt-contrib-less.

**Comment by joshdmiller** [comment]
Thanks. I took everyone's advice and switched to grunt-contrib-less for
version 0.4. It's a major overhaul, so I haven't pushed the changes yet.
I'll be looking for beta testers next week; anyone interested?

Back to this - I wouldn't object to pushing a patch release to 0.3 if this
is an important issue, in which case a PR would be awesome. :-)

**Comment by alex6o** [comment]
thx for switching :) btw ... do you already have a release date for 0.4?

**Comment by joshdmiller** [comment]
Dude, that is an ugly question. :-)

The release date is a month ago. So, yeah... October was a hugely busy month for me; November is a little lighter, but I don't have a lot of free time. I'm hoping to release a preliminary end of next week, but that could easily change.

Version 0.4 will be _very_ different. Optional components (like less, sass, coffeescript, etc.) will not be included by default, but will be easily available; there will be a command line tool for interacting with ngbp; `build.config.js` and `Gruntfile.js` will be significantly reduced and combined; the ngbp functionality will be put into an NPM package to make updates easy as pie; ngbp will become a more generic build system allowing much more awesome, powerful, and simpler configuration; the documentation will be more comprehensive, with tutorials; I will be making screencasts to help with the transition; I'm mulling adding generators like Yeoman, but have mixed feelings here; et cetera...

So it's a _huge_ overhaul. In fact, this one repository will be split into several ones and this repository will be the "demo" of how to use it (with bootstrap and less, probably). Currently, I've got the new build system working and am in the process of splitting all the tasks into their own modules. This is sort of time-consuming, but not as bad as it probably sounds.

That's why I need beta testers. If I can get people interested, I'll host a hangout explaining all the changes and get people ready to try it out. Hopefully, some of those will be interested in making contributions so I can expand the team because, long-term, this is too much for one person.

So anyway, if I got a free day or two I could get it done. Instead, I'm having to work it in here and there - I'm sure you can relate. Anyway, November for sure. Yeah, November. Probably. ... Maybe. :-/

**Comment by PaulL1** [comment]
I'm interested in participating, I'm updating my tutorial that made heavy use of ng-boilerplate to deal with Rails 4, Bootstrap 3 and some changes to the way I'd built things.  I'd be keen to pilot your new setup as part of that, given that it's a demo app it doesn't come with fixed timelines, and it's simple enough to refactor if the direction changes without losing months of client chargeable work.  :-)

In terms of features I'd love to see as part of that (and that I was missing in the prior version):
- Merging of the end to end testing.  I was kinda waiting for that merge, but had on my list for this week to merge it myself.  If 4.0 is coming instead, that'd be better
- Some sort of distinction between module level testing and end-to-end testing.  With the way ng-boilerplate works I'd like to create 
  - unit tests (in the traditional sense of that phrase - so mocks and harnesses, and testing each controller method for each branch and loop)
  - module tests.  This would test that all the things that are together in a folder actually work together.  Given the ui-state setup, and the reasonably clean handling of dependencies that ng-boilerplate encourages, I'd like to have tests within my module directory that test the template + the controller + any filters or other logic that I have - basically that if you navigate to the state that is described in this module all the pages inside the module look as they should.  I think this still wouldn't interact with the server - so I'd still be mocking $http
  - module e2e tests.  This would test the module + the server together
  - e2e tests.  This would run the application from end to end, chaining a series of modules together
- some of the dependency generation/inclusion stuff, perhaps from #179
- making sure that when you pull down ng-boilerplate (from npm or from elsewhere) that it runs straight out of box.  It does run now, but gives a couple of errors around fonts, it'd be nice to merge in the fixes for that

In short, ready and willing to contribute and test.

**Comment by PaulL1** [comment]
Actually, having written my long list, I've also spent a bit of time working through how I'd do this.  I think this isn't something that needs to be in the framework, rather each of the tests I describe is either an e2e test or a unit test in ng-boilerplate terms, and as such I don't need extra functionality to do this, I just need to get organised.  So, for now I'm going to try to merge the e2e testing changes and see whether I can build tests that way.

**Comment by joshdmiller** [comment]
@PaulL1 Interesting thoughts.

> Merging of the end to end testing. I was kinda waiting for that merge, but had on my list for this week to merge it myself. If 4.0 is coming instead, that'd be better

I definitely agree that e2e testing is a must in 0.4. It may not be in next week's pre-beta, but it will definitely be in the final release this month.

> Some sort of distinction between module level testing and end-to-end testing.

I'm not sure how this would be implemented. Can you give an example?

> some of the dependency generation/inclusion stuff, perhaps from #179

I agree. I was on the fence about this before, but I'm on board now. Something like #179 will be included as part of the LESS and SASS plugins for NGBP.

> making sure that when you pull down ng-boilerplate (from npm or from elsewhere) that it runs straight out of box. It does run now, but gives a couple of errors around fonts, it'd be nice to merge in the fixes for that

This is already addressed in 0.4. There is a distinction between the boilerplate (which is grunt-installable) and the example code (which has Bootstrap and fonts). But the demo will run without errors due to font inclusion changes.

Speaking of the pre-beta, I'm going to host a little Google+ hangout maybe at the end of next week with whomever is interested in beta testing 0.4. I don't have documentation up yet (I anticipate some significant changes prior to the release) so I will show everyone how to use it, answer questions, etc. It would be awesome if y'all could join.

**Comment by PaulL1** [comment]
On module level and end-2-end, I'm probably drawing two distinctions.  The first is whether I'm mocking my server calls.  In end-2-end testing I wouldn't be, in at least some of my module level I would be.  So in a sense my module level testing is unit testing + checking that my page is actually rendering.

To give you an example, I have pretty extensive unit testing in my app at the moment.  I have a breadcrumb service that renders breadcrumbs at the top of the page.  I refactored that service and moved it to a different module, and fixed all the unit tests.  I reached a point where all my unit tests were passing again (I had to go through each one changing the name of the breadcrumb service), so I was happy.  But when I went to the UI, no breadcrumbs.  Because of course the html page template binds directly to the variable name, so it wasn't showing them any more.  

What I'm looking for at module level testing is that I verify that everything from my angular model is actually showing in the right places on my html page.  I'm doing i18n using angular-translate, so another module level test thing for me is to verify that all the translations on that page actually run (at least for one of the languages, probably not for all of them, although if I can make that verification generic enough, I could arguably check every language).

End-2-end testing I see as being more functional in nature, and therefore to verify fewer error situations - so I wouldn't really see my end-2-end testing verifying that each individual field has the data I expect in it, or that each individual column heading on my grid had been translated.  

Does that make sense, or am I perhaps thinking about it wrong?  I did see mention on this blog post (http://www.yearofmoo.com/2013/01/full-spectrum-testing-with-angularjs-and-karma.html) of a concept called "midway testing".  I'm not yet convinced that it's the same thing I'm looking for - I think the ng-boilerplate framework encourages us to think about things a bit differently.  But it is perhaps aiming to resolve the same problem.

Definitely interested in the google hangout.  I'm in Australia, and have a day job, but so long as it's a time I can hit then I'll come.

**Comment by PaulL1** [comment]
Another thought.  Looking at the angular end to end testing page, it says "If you're starting a new Angular project, you may want to look into using Protractor, as it is going to replace the current method of E2E Testing in the near future."  http://docs.angularjs.org/guide/dev_guide.e2e-testing

Given you're doing a rewrite, it probably makes sense to take note of that.  I'm about to start working out what protractor does and how, so I may be able to provide some information once I've done that.

**Comment by joshdmiller** [comment]
Protractor has has been on my list of things to mess with for a while; I'm
only superficially familiar with it from a few minutes of messing around -
perhaps that _would_ be a good idea, Paul...

**Comment by lanoxx** [comment]
@rickhuizinga 
Have you uploaded your changes somewhere, where I could have a look at them. I also have this problem and a back port or at least some references of how it can be done would be very helpful. Currently my theme is partially broken because of this issue and I havent found a workaround yet.

Does this stand any chance of being back ported to 0.3.x?

**Comment by joshdmiller** [comment]
@lanoxx Absolutely. With a pull request, I'd be happy to merge it to the 3.x branch.

**Comment by iulianmihai** [comment]
For everyone who is interested, I found the following solution:

``` less
height: ~"(calc(~'100%' - ~'116px'))";
```

output:

``` less
height: calc(100% - 116px);
```

The reason I'm not using  grunt-contrib-less is because I can not import less files in other less files.

**Comment by kbatra0108** [comment]
HI,
I am getting "ParseError: directive options not recognised" on compiling the line below:

@include transition(all 0.30s ease-in-out);

Help me out please !!!

**Comment by lanoxx** [comment]
For me this issue was resolved by the switch from twitter-recess to grunt-contrib-less and since that patch is included in 0.3.2 I think we can probably close this issue.

@kbatra0108 : I guess there is a problem with your expression and not a problem of escaping.

@iulianmihai : Can you better explain what you mean by not being able to import less files? Isn't that just possible with the `@import` directive?

**Comment by iulianmihai** [comment]
@lanoxx I think I was wrong about grunt-contrib-less I wanted to refer to grunt-recess:
The following line will not work...

``` less
@import (less) "/bower_components/mycomponent/dist/less/mycomponent.css";
```

https://github.com/twitter/recess/issues/121

**Comment by PawelOwczarekFalcon** [comment]
I found solution:
http://gillianshelpdesk.tumblr.com/post/127773807047/calc-issues-in-grunt-contrib-less

**Comment by Marlywoos** [comment]
@iulianmihai  Your solution is perfect! Tks
